source/common/http/filter_manager.h

```diff
@@ -10,22 +10,107 @@
 #include "common/grpc/common.h"
 #include "common/http/headers.h"
 #include "common/local_reply/local_reply.h"
+#include "common/matcher/matcher.h"
 
 namespace Envoy {
 namespace Http {
 
 class FilterManager;
 
+struct HttpMatchingData : public MatchingData, public MatchTreeFactoryCallbacks {
+public:
+  // MatchTreeFactoryCallbacks
+  void addPredicateMatcher(MatchWrapperSharedPtr matcher) {
+    matchers_.push_back(std::move(matcher));
+  }
+
+  void onNewStream() {
+    for (const auto& matcher : matchers_) {
+      matcher->rootMatcher().onNewStream(matcher->status_);
+    }
+  }
+
+  void onRequestHeaders(Http::RequestHeaderMap& request_headers) {
+    request_headers_ = &request_headers;
+    for (const auto& matcher : matchers_) {
+      matcher->rootMatcher().onHttpRequestHeaders(request_headers, matcher->status_);
+    }
+  }
+
+  void onRequestData(const Buffer::Instance& buffer) {
+    for (const auto& matcher : matchers_) {
+      matcher->rootMatcher().onRequestBody(buffer, matcher->status_);
+    }
+  }
+
+  void onRequestTrailers(Http::RequestTrailerMap& request_trailers) {
+    request_trailers_ = &request_trailers;
+    for (const auto& matcher : matchers_) {
+      matcher->rootMatcher().onHttpRequestTrailers(request_trailers, matcher->status_);
+    }
+  }
+
+  void onResponseHeaders(Http::ResponseHeaderMap& response_headers) {
+    response_headers_ = &response_headers;
+    for (const auto& matcher : matchers_) {
+      matcher->rootMatcher().onHttpResponseHeaders(response_headers, matcher->status_);
+    }
+  }
+
+  void onResponseData(const Buffer::Instance& buffer) {
+    for (const auto& matcher : matchers_) {
+      matcher->rootMatcher().onResponseBody(buffer, matcher->status_);
+    }
+  }
+
+  void onResponseTrailers(Http::ResponseTrailerMap& response_trailers) {
+    response_trailers_ = &response_trailers;
+    for (const auto& matcher : matchers_) {
+      matcher->rootMatcher().onHttpResponseTrailers(response_trailers, matcher->status_);
+    }
+  }
+
+  std::vector<MatchWrapperSharedPtr> matchers_;
+  Http::RequestHeaderMap* request_headers_;
+  Http::RequestTrailerMap* request_trailers_;
+  Http::ResponseHeaderMap* response_headers_;
+  Http::ResponseTrailerMap* response_trailers_;
+};
+using HttpMatchingDataSharedPtr = std::unique_ptr<HttpMatchingData>;
+
+class HttpKeyNamespaceMapper : public KeyNamespaceMapper {
+public:
+  void forEachValue(absl::string_view ns, absl::string_view key, const MatchingData& matching_data,
+                    std::function<void(absl::string_view)> value_cb) override {
+    const HttpMatchingData& http_data = dynamic_cast<const HttpMatchingData&>(matching_data);
+    // TODO(snowp): We only support request headers for now, add others.
+    if (ns == "request_headers") {
```

Comment:
Seems like this could be pre-converted somehow to both check that it's valid and all a faster check against an enum?